feat(curriculum): add case converter project to teach list comprehension (#52569)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-list-comprehension-by-building-a-case-converter-program/657f43d341a0dd17120cdb08.md

---
id: 657f43d341a0dd17120cdb08
title: Step 17
challengeType: 20
dashedName: step-17
---

# --description--

> Inside the space you left between the pair of square braces, you can describe your desired value based on a given condition.

Reviewer: ```suggestion
Inside the space you left between the pair of square braces, you can describe the value that you would like to include in the list based on a given condition.
```

```py
snake_cased_char_list = [
    '_' + char.lower() if char.isupper()
]
```

> Python will interpret this expression as "append `'_' + char.lower()` to the list if `char` is in uppercase" and this covers the case for the capital letters in the input string.

Reviewer: ```suggestion
Python will interpret this expression as "append `'_' + char.lower()` to the list if `char` is in uppercase" and this will convert the case for the capital letters in the input string.
```

Start by adding this line within the square braces.

# --hints--

Test 1

```js

```

# --seed--

## --seed-contents--

```py
def convert_to_snake_case(pascal_or_camel_cased_string):
    # snake_cased_char_list = []
    # for char in pascal_or_camel_cased_string:
    #     if char.isupper():
    #         snake_cased_char_list.append(f'_{char.lower()}')
    #     else:
    #         snake_cased_char_list.append(char)
    # snake_cased_string = ''.join(snake_cased_char_list)
    # clean_snake_cased_string = snake_cased_string.strip('_')

    # return clean_snake_cased_string

>     snake_cased_char_list = [
> --fcc-editable-region--
> 
> 
> 
> --fcc-editable-region--
>     ]

Reviewer 2: ```suggestion
--fcc-editable-region--
    snake_cased_char_list = [
        
    ]
--fcc-editable-region--
```

> 

Reviewer: I would suggest removing one blank line to keep only two blank lines.

Author: Done

---
id: 657f43d341a0dd17120cdb08
title: Step 17
challengeType: 20
dashedName: step-17
---

# --description--

Inside the space you left between the pair of square braces, you can describe the value that you would like to include in the list based on a given condition.

```py
snake_cased_char_list = [
    '_' + char.lower() if char.isupper()
]
```

Python will interpret this expression as "append `'_' + char.lower()` to the list if `char` is in uppercase" and this will convert the case for the capital letters in the input string.

Start by adding this line within the square braces.

# --hints--

You should add `'_' + char.lower() if char.isupper()` within the square braces of `snake_cased_char_list`.

```js
const transformedCode = e.code.original["main.py"].replace(/\r/g, "");
const convert_to_snake_case = __helpers.python.getDef("\n" + transformedCode, "convert_to_snake_case");
const { function_body } = convert_to_snake_case;

> assert.match(function_body, / +snake_cased_char_list\s*=\s*\[\n*\s*'_'\s*\+\s*char.lower\(\)\s+if\s+char.isupper\(\s*\)\n*\]/);

Reviewer 2: ```suggestion
assert.match(function_body, / +snake_cased_char_list\s*=\s*\[\s*("|')_\1\s*\+\s*char\.lower\(\s*\)\s+if\s+char\.isupper\(\s*\)\s*\]/);
```

---
id: 657f43d341a0dd17120cdb08
title: Step 17
challengeType: 20
dashedName: step-17
---

# --description--

Inside the space you left between the pair of square braces, you can describe the value that you would like to include in the list based on a given condition.

```py
snake_cased_char_list = [
    '_' + char.lower() if char.isupper()
]
```

Python will interpret this expression as "append `'_' + char.lower()` to the list if `char` is in uppercase" and this will convert the case for the capital letters in the input string.

Start by adding this line within the square braces.

# --hints--

You should add `'_' + char.lower() if char.isupper()` within the square braces of `snake_cased_char_list`.

```js

> ({
>     test: () => {
>         const transformedCode = e.code.original["main.py"].replace(/\r/g, "");
>         const convert_to_snake_case = __helpers.python.getDef("\n" + transformedCode, "convert_to_snake_case");
>         const { function_body } = convert_to_snake_case;
> 
>         assert.match(function_body, / +snake_cased_char_list\s*=\s*\[\n*\s*'_'\s*\+\s*char.lower\(\)\s{1}if\s{1}char.isupper\(\)\n*\]/);
>     }
> })

Reviewer 2: ```suggestion
const transformedCode = e.code.original["main.py"].replace(/\r/g, "");
const convert_to_snake_case = __helpers.python.getDef("\n" + transformedCode, "convert_to_snake_case");
const { function_body } = convert_to_snake_case;

assert.match(function_body, / +snake_cased_char_list\s*=\s*\[\n*\s*'_'\s*\+\s*char.lower\(\)\s+if\s+char.isupper\(\s*\)\n*\]/);
```

Reviewer 2: Since the code is incomplete it throws an error and that test format is not suitable.
